cedar-policy/src/api.rs

```diff
@@ -3386,9 +3383,115 @@ impl<'a> RequestBuilder<'a> {
 
     /// Set the schema. If present, this will be used for request validation.
     #[must_use]
-    pub fn schema(self, schema: &'a Schema) -> Self {
+    pub fn schema(self, schema: &Schema) -> ValidatedRequestBuilder<'_> {
+        ValidatedRequestBuilder {
+            principal: self.principal,
+            action: self.action,
+            resource: self.resource,
+            context: self.context,
+            schema,
+        }
+    }
+
+    /// Create the [`Request`]
+    pub fn build(self) -> Request {
+        Request(ast::Request::new_unchecked(
+            self.principal,
+            self.action,
+            self.resource,
+            self.context,
+        ))
+    }
+}
+
+/// Schema validated builder for a [`Request`]
+#[cfg(feature = "partial-eval")]
+#[derive(Debug)]
+pub struct ValidatedRequestBuilder<'a> {
+    principal: ast::EntityUIDEntry,
+    action: ast::EntityUIDEntry,
+    resource: ast::EntityUIDEntry,
+    /// Here, `None` means unknown
+    context: Option<ast::Context>,
+    schema: &'a Schema,
+}
+
+#[cfg(feature = "partial-eval")]
+impl<'a> ValidatedRequestBuilder<'a> {
+    /// Set the principal.
+    ///
+    /// Note that you can create the `EntityUid` using `.parse()` on any
+    /// string (via the `FromStr` implementation for `EntityUid`).
+    ///
+    /// Here, passing `None` for `principal` indicates that `principal` does
+    /// not contribute to authorization decisions (e.g., because it is not
+    /// used in your policies).
+    /// This is different than Unknown for partial-evaluation purposes.
+    #[must_use]
+    pub fn principal(self, principal: Option<EntityUid>) -> Self {
```

Comment:
It's a bit of a pity to have to duplicate this code & documentation from `RequestBuilder`, but I don't have a better idea atm so approving as-is.

Reply:
I have tweaked things a bit to avoid this (and avoided the whole `ValidatedRequestBuilder` thing altogether :)